gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+                        filesToBeDeleted);
+                outputRows.add(outputRow);
+              });
+
+      dropExpireProcedureBranch(expiryBranchName, api);
+
+      GCCheckPointRepo gcCheckPointRepo =
+          new GCCheckPointRepo(
+              spark(), gcCatalogName, gcOutputBranchName, gcCheckPointTableIdentifier);
+      Row markerRow = gcCheckPointRepo.createMarkerRow(runId);
+      gcCheckPointRepo.writeToOutputTable(Collections.singletonList(markerRow));
+    }
+    return outputRows.toArray(new InternalRow[0]);
+  }
+
+  private static void addOneContentToExpireProcedureBranch(
+      String expiryBranchName,
+      IdentifiedResultsRepo identifiedResultsRepo,
+      NessieApiV1 api,
+      String finalRunId,
+      String contentId) {
+    String commitHash =
+        identifiedResultsRepo.getAnyCommitHashForContentId(contentId, finalRunId).get();
+    Content content = getContentFromCommitHash(api, commitHash, contentId);
+    // keep the content key same as content id with dummy implicit namespace
+    // EXPIRE_NAMESPACE
+    String headCommitHash = getHeadCommitHash(api, expiryBranchName);
+    CommitMeta commitMeta = CommitMeta.builder().message("putting the contents for GC").build();
+    try {
+      api.commitMultipleOperations()
+          .branch(Branch.of(expiryBranchName, headCommitHash))
+          .commitMeta(commitMeta)
+          .operation(Operation.Put.of(ContentKey.of(EXPIRE_NAMESPACE, contentId), content, content))
+          .commit();
+    } catch (NessieNotFoundException | NessieConflictException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static ExpireSnapshots.Result executeExpireSnapshotsPerTable(
+      boolean dryRun,
+      Map.Entry<String, List<Long>> entry,
+      List<String> filesToBeDeleted,
+      Table table) {
+    Consumer<String> expiredFilesConsumer =
+        getConsumerBasedOnDryRunConfig(dryRun, filesToBeDeleted, (HasTableOperations) table);
+    ExpireSnapshots expireSnapshots =
+        SparkActions.get().expireSnapshots(table).deleteWith(expiredFilesConsumer);
+    entry.getValue().forEach(expireSnapshots::expireSnapshotId);
+    return expireSnapshots.execute();
+  }
+
+  private static Consumer<String> getConsumerBasedOnDryRunConfig(
+      boolean dryRun, List<String> filesToBeDeleted, HasTableOperations table) {
+    Consumer<String> expiredFilesConsumer;
+    if (!dryRun) {
+      expiredFilesConsumer =
+          fileName -> {
+            filesToBeDeleted.add(fileName);
+            // delete the expired files
+            table.operations().io().deleteFile(fileName);
+          };
+    } else {
+      expiredFilesConsumer = filesToBeDeleted::add;
+    }
+    return expiredFilesConsumer;
+  }
+
+  private static void updateMetadataLocation(BaseTable table) {
+    try {
+      BaseMetastoreTableOperations operations = (BaseMetastoreTableOperations) table.operations();
+      FieldUtils.writeField(
```

Comment:
Same strong -1 regarding reflection here, too

Reply:
this is temporary, as mentioned in the method caller, I fixed this bug at Iceberg side. I can remove it once Iceberg makes a release.